modules/common-fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/common/fs2/Enrich.scala

```diff
@@ -64,33 +68,40 @@ object Enrich {
    * [[Assets.State.make]] downloads assets for the first time unconditionally during
    * [[Environment]] initialisation, then if `assetsUpdatePeriod` has been specified -
    * they'll be refreshed periodically by [[Assets.updateStream]]
-   *
-   * @param ordered indicates whether the events should be processed ordered or not
    */
-  def run[F[_]: Concurrent: ContextShift: Clock: Parallel, A](env: Environment[F, A], ordered: Boolean): Stream[F, Unit] = {
+  def run[F[_]: Concurrent: ContextShift: Clock: Parallel: Timer, A](env: Environment[F, A]): Stream[F, Unit] = {
     val registry: F[EnrichmentRegistry[F]] = env.enrichments.get.map(_.registry)
     val enrich: Enrich[F] = {
       implicit val rl: RegistryLookup[F] = env.registryLookup
       enrichWith[F](registry, env.igluClient, env.sentry, env.processor)
     }
 
-    val enrichPipe: Pipe[F, (A, Array[Byte]), (A, Result)] =
-      in => {
-        val parallel =
-          if (ordered)
-            in.parEvalMap[F, (A, Result)](env.streamsSettings.concurrency.enrichment) _
-          else
-            in.parEvalMapUnordered[F, (A, Result)](env.streamsSettings.concurrency.enrichment) _
-        parallel { case (orig, bytes) => enrich(bytes).map((orig, _)) }
-      }
-
-    env.source
-      .pauseWhen(env.pauseEnrich)
-      .evalTap(_ => env.metrics.rawCount)
-      .fproduct(env.getPayload)
-      .through(enrichPipe)
-      .through(sinkResult(sinkOne(env), env.metrics.enrichLatency, env.streamsSettings.concurrency.output))
-      .through(env.checkpointer)
+    val enriched =
+      env.source.chunks
+        .prefetch
+        .evalTap(chunk => Logger[F].debug(s"${DateTime.now().toString()} - Starting to process chunk of size ${chunk.size}"))
+        .pauseWhen(env.pauseEnrich)
+        .evalTap(chunk => chunk.toList.map(_ => env.metrics.rawCount).sequence_)
+        .map(chunk => chunk.map(a => (a, env.getPayload(a))))
+        .evalTap(chunk => Logger[F].debug(s"${DateTime.now().toString()} - Starting to enrich chunk of size ${chunk.size}"))
+        .evalMap(chunk =>
+          chunk.toList.map { case (orig, bytes) => enrich(bytes).map((orig, _)) }.parSequenceN(env.streamsSettings.concurrency.enrich)
+        )
+        .prefetch
+
+    val sinkAndCheckpoint: Pipe[F, List[(A, Result)], Unit] =
+      _
+        .evalTap(chunk => Logger[F].debug(s"${DateTime.now().toString()} - Starting to sink chunk of size ${chunk.size}"))
+        .parEvalMap(env.streamsSettings.concurrency.sink) { chunk =>
+          val records = chunk.map(_._1)
+          val results = chunk.map(_._2)
+          sinkChunk(results, sinkOne(env), env.metrics.enrichLatency).as(records)
+        }
+        .prefetch
+        .evalTap(chunk => Logger[F].debug(s"${DateTime.now().toString()} - Starting to checkpoint chunk of size ${chunk.size}"))
+        .evalMap(env.checkpoint)
```

Comment:
The old version of stream enrich checkpointed whenever [timeLimit](https://github.com/snowplow/enrich/blob/2.0.3/config/config.hocon.sample#L184) is reached.  There was no default, but in open source terraform modules [the default is 500 ms](https://github.com/snowplow-devops/terraform-aws-enrich-kinesis-ec2/blob/1ca899400ff159db2fbf4fc0c7fcc64ba2736881/variables.tf#L142).

Whereas the new enrich checkpoints after every batch, which by default is [10000 records](https://github.com/snowplow/enrich/blob/5778418d5ccf8ace73844cf5ed0110b768e3407c/modules/kinesis/src/main/resources/application.conf#L8).

I'm think I'm OK with this.  But I might as well tag @jbeemster who might have a comment on that!

Reply:
`timeLimit` is not the checkpoint interval, but the maximum duration to which events are accumulated in a buffer before being sent to Kinesis. I wonder if checkpointing is triggered each time a buffer is sent to Kinesis.

With our design, we can't checkpoint more often, as we checkpoint records as soon as they have been sent to Kinesis.

We could find a way to checkpoint less often, but we are already checkpointing slower than once per minute, it should be totally fine with DynamoDB quotas.